logging/logging.go

```diff
@@ -24,77 +24,253 @@ package logging
+		TimeKey:        "ts",
+		LevelKey:       "level",
+		NameKey:        "name",
+		CallerKey:      "caller",
+		MessageKey:     "message",
+		StacktraceKey:  "stacktrace",
+		LineEnding:     zapcore.DefaultLineEnding,
+		EncodeLevel:    zapcore.CapitalLevelEncoder,
+		EncodeTime:     zapcore.ISO8601TimeEncoder,
+		EncodeDuration: zapcore.StringDurationEncoder,
+		EncodeCaller:   zapcore.ShortCallerEncoder,
+	}
+}
+
+func NewConfig() zap.Config {
+	return zap.Config{
+		Level:            zap.NewAtomicLevelAt(zap.DebugLevel),
+		Development:      false,
+		Encoding:         "console",
+		EncoderConfig:    NewEncoderConfig(),
+		OutputPaths:      []string{"stderr"},
+		ErrorOutputPaths: []string{"stderr"},
+	}
+}
+
+func initLogger() (err error) {
+	cfg := config.GetConfig()
+	backendLevel := getZapLevel(cfg.GetString("logging.level"))
+	consoleEncoder := zapcore.NewConsoleEncoder(NewEncoderConfig())
+	jsonEncoder := zapcore.NewJSONEncoder(NewEncoderConfig())
+	msgPriority := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return lvl >= backendLevel
+	})
+
+	var backends []zapcore.Core
 	for _, name := range cfg.GetStringSlice("logging.backends") {
 		switch name {
 		case "file":
 			filename := cfg.GetString("logging.file.path")
 			file, err := os.Create(filename)
 			if err != nil {
-				return nil, err
+				return err
 			}
-			backend = logging.NewLogBackend(file, "", 0)
+			backends = append(backends, zapcore.NewCore(consoleEncoder, zapcore.Lock(file), msgPriority))
 		case "syslog":
-			backend, err = logging.NewSyslogBackend("")
+			w, err := syslog.New(syslog.LOG_CRIT, "")
 			if err != nil {
-				return nil, err
+				return err
 			}
+			backends = append(backends, zapcore.NewCore(consoleEncoder, zapcore.AddSync(w), msgPriority))
 		case "stderr":
-			backend = logging.NewLogBackend(os.Stderr, "", 0)
+			backends = append(backends, zapcore.NewCore(consoleEncoder, zapcore.Lock(os.Stderr), msgPriority))
+		case "json":
+			backends = append(backends, zapcore.NewCore(jsonEncoder, zapcore.Lock(os.Stdout), msgPriority))
 		default:
-			return nil, fmt.Errorf("Invalid logging backend: %s", name)
+			return fmt.Errorf("Invalid logging backend: %s", name)
 		}
-
-		backend = logging.NewBackendFormatter(backend, logging.MustStringFormatter(format))
-		backends = append(backends, backend)
 	}
 
-	backendLevel := logging.MultiLogger(backends...)
-	backendLevel.SetLevel(level, "")
+	newCore := zap.WrapCore(func(zapcore.Core) zapcore.Core {
+		return zapcore.NewTee(backends...)
+	})
 
-	logger, err := logging.GetLogger("")
-	if err != nil {
-		return nil, err
+	c := NewConfig()
+	c.Level.SetLevel(backendLevel)
+	z, _ := c.Build(
+		newCore,
+		zap.AddCallerSkip(2),
+		// uncomment the following line to get stacktrace on error messages
+		// zap.AddStacktrace(zapcore.ErrorLevel),
+		zap.AddStacktrace(zapcore.DPanicLevel),
+	)
+	logger = &Logger{
+		Logger: z,
+		config: c,
+		id:     cfg.GetString("host_id") + ":" + cfg.GetString("logging.id"),
```

Comment:
my concern was not the cost but the output which will have a ':' with nothing following